### PR TITLE
Create hash_map_open_addressing.c

### PR DESCRIPTION
**Pull request by lclc6**
翻译哈希冲突中C语言相关代码

If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

 I've tested the code and ensured the outputs are the same as the outputs of reference codes.
 I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
 The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 1]
```diff
@@ -0,0 +1,196 @@
+/**
+ * File: hash_map_open_addressing.c
+ * Created Time: 2023-10-6
+ * Author: lclc6 (w1929522410@163.com)
+ */
+
+#include "../utils/common.h"
+
+/* 使用开发地址法来解决哈希冲突*/
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:9]: Unnecessary

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+/* 开放寻址哈希表 */
+struct Pair{
+    int key;
+    char *val;
+};
+typedef struct Pair Pair;
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:16]: Leave a blank line in the middle

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    hashMap->TOMBSTONE = (Pair *)malloc(sizeof(Pair));
+    hashMap->TOMBSTONE->key = -1;
+    hashMap->TOMBSTONE->val = "-1";
+
+    return hashMap;
+}
+
+/* 析构方法 */
+void delHashMapOpenAddressing(HashMapOpenAddressing *hashMap) {
+    for(int i = 0;i < hashMap->capacity;i++) {
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:46]: `for (`

Comment (Gonglja) [reply codes/c/chapter_hashing/hash_map_open_addressing.c:46]: `; `

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        Pair *pair = hashMap->buckets[i];
+        if(pair == NULL || pair->key == -1) continue;
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:48]: `if (`

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        free(pair->val);
+    }
+    free(hashMap->TOMBSTONE);
+    free(hashMap);
+
+    return ;
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:54]: Unnecessary

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+}
+
+/* 哈希函数 */
+int hashFunc(HashMapOpenAddressing *hashMap,int key) {
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:58]: Pay attention to the space between the two formal parameters

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    return key % hashMap->capacity;
+}
+
+/* 负载因子 */
+double loadFactor(HashMapOpenAddressing *hashMap) {
+    return (double)hashMap->size / hashMap->capacity;
+}
+
+/* 搜索 key 对应的桶索引 */
+int findBucket(HashMapOpenAddressing *hashMap,int key) {
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:68]: Same as above

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    int index = hashFunc(hashMap,key);
+    int firstTombstone = -1;
+    // 线性探测，当遇到空桶时跳出
+    while(hashMap->buckets[index] != NULL) {
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:72]: `while `

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+            printf("NULL\n");
+        }else if(pair == hashMap->TOMBSTONE) {
+            printf("TOMBSTONE\n");
+        }else {
+            printf("%d -> %s\n",pair->key,pair->val);
+        }
+    }
+}
+
+int main(int argc, char const *argv[])
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:184]: /* Driver Code */

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+{
+    HashMapOpenAddressing *map = newHashMapOpenAddressing();
+    put(map,0,"0000");
+    put(map,5,"hh");
+    put(map,1,"lc");
+    put(map,9,"hello");
+    removeItem(map,5);
+    print(map);
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:192]: Test cases should be consistent with other code

**Review by Gonglja — COMMENTED** [review]
Pay attention to the format of the code. There are no spaces after a large number of `if` and `while` keywords in the text.
It is recommended to use vscode to directly format the code. 
For details, see https://www.hello-algo.com/chapter_appendix/installation/#3-cc

File: codes/c/chapter_hashing/hash_map_open_addressing.c (hunk 10 above)
Comment (lclc6) [reply codes/c/chapter_hashing/hash_map_open_addressing.c]: 代码已修改

Comment (lclc6) [reply codes/c/chapter_hashing/hash_map_open_addressing.c]: > e



> Test cases should be consistent with other code

代码已修改

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+        } else if (pair == hashMap->TOMBSTONE) {
+            printf("TOMBSTONE\n");
+        } else {
+            printf("%d -> %s\n", pair->key, pair->val);
+        }
+    }
+}
+
+/* Driver Code */
+int main(int argc, char const *argv[]) {
```
Comment (Gonglja) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:186]: The main function does not need to pass parameters for the time being

**Review by Gonglja — COMMENTED** [review]
@krahets see if you have any other questions?

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+/* 开放寻址哈希表 */
+struct Pair {
+    int key;
+    char *val;
+};
+
+typedef struct Pair Pair;
+
+struct HashMapOpenAddressing {
```
Comment (krahets) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:17]: I noticed that some structs are named in camelCase, while others are PascalCase. 

Which one is more popular? @Gonglja

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+                hashMap->buckets[index] = hashMap->TOMBSTONE;
+                return firstTombstone; // 返回移动后的桶索引
+            }
+            return index; // 返回桶索引
+        }
+        // 记录遇到的首个删除标记
+        if (firstTombstone == -1 && hashMap->buckets[index] == hashMap->TOMBSTONE) {
+            firstTombstone = index;
+        }
+        // 计算桶索引，越过尾部返回o头部
```
Comment (krahets) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:88]: 返回o头部

File: codes/c/chapter_hashing/hash_map_open_addressing.c [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+/* 析构方法 */
+void delHashMapOpenAddressing(HashMapOpenAddressing *hashMap) {
+    for (int i = 0; i < hashMap->capacity; i++) {
+        Pair *pair = hashMap->buckets[i];
+        if (pair == NULL || pair->key == -1)
+            continue;
+        free(pair->val);
+        free(pair);
+    }
+    return;
```
Comment (krahets) [inline codes/c/chapter_hashing/hash_map_open_addressing.c:55]: redundant line?

**Review by krahets — APPROVED** [review]
Thanks for the clean code! Look forward to your next PR

**Pull request merged**